kernel_module/legion-laptop.c

```diff
@@ -2842,84 +2875,93 @@ struct WMIFanTableRead {
 	u32 FSS7;
 	u32 FSS8;
 	u32 FSS9;
-	u32 FSSA;
+	// Ignored for now
+	u32 FTLE;
+	u32 FST0;
+	u32 FST1;
+	u32 FST2;
+	u32 FST3;
+	u32 FST4;
+	u32 FST5;
+	u32 FST6;
+	u32 FST7;
+	u32 FST8;
+	u32 FST9;
 } __packed;
 
 static ssize_t wmi_read_fancurve_custom(const struct model_config *model,
 					struct fancurve *fancurve)
 {
-	u8 buffer[88];
+	struct WMIFanTableRead fantable;
 	int err;
 
-	// The output buffer from the ACPI call is 88 bytes and larger
-	// than the returned object
-	pr_info("Size of object: %lu\n", sizeof(struct WMIFanTableRead));
-	err = wmi_exec_noarg_ints(WMI_GUID_LENOVO_FAN_METHOD, 0,
-				  WMI_METHOD_ID_FAN_GET_TABLE, buffer,
-				  sizeof(buffer));
+	struct acpi_buffer params;
+	u32 fanid = 0;
+	params.length = sizeof(fanid);
+	params.pointer = &fanid;
+
+	err = wmi_exec_ints(WMI_GUID_LENOVO_FAN_METHOD, 0,
+			    WMI_METHOD_ID_FAN_GET_TABLE, &params, &fantable,
+			    sizeof(fantable));
 	print_hex_dump(KERN_INFO, "legion_laptop fan table wmi buffer",
-		       DUMP_PREFIX_ADDRESS, 16, 1, buffer, sizeof(buffer),
-		       true);
+		       DUMP_PREFIX_ADDRESS, 16, 1, &fantable,
+		       sizeof(fantable), true);
 	if (!err) {
-		struct WMIFanTableRead *fantable =
-			(struct WMIFanTableRead *)&buffer[0];
 		fancurve->current_point_i = 0;
-		fancurve->size = 10;
+		fancurve->size = fantable.FSTL;
```

Comment:
Hmm. This does not work for other models because FSTL is not set to 10 but just not set.